refactor: Use deduction guide for std::array<uint8_t, N>

hebasto opened the pull request:
> Using the C++11 std::array with explicit template parameters is problematic because overshooting the size will fill the memory with default constructed types.

[Credits](https://github.com/bitcoin/bitcoin/pull/20566) to **MarcoFalke**.

hebasto committed:
refactor: Use deduction guide for std::array<uint8_t, N>

practicalswift commented:
Concept ACK

elichai commented on src/util/string.h:
```diff
@@ -87,4 +87,9 @@ template <typename T1, size_t PREFIX_LEN>
+/**
+ * User-defined literal for byte constants.
+ */
+constexpr uint8_t operator ""_u8(unsigned long long byte) { return byte; }
```
Maybe we should add some overflow assert here? it would be nice if `256_u8` won't compile.
the downside is that this function might be used outside of constexpr, but that's very unlikely

elichai replied:
ie `assert(byte <= std::numeric_limits<std::uint8_t>::max());`

hebasto replied:
> Maybe we should add some overflow assert here? it would be nice if `256_u8` won't compile.

Agree...

> ie `assert(byte <= std::numeric_limits<uint8_t>::max());`

... but this won't work, unfortunately.

elichai replied:
Why not? 
https://godbolt.org/z/hTjch7

hebasto replied:
It is not a compile-time error.

elichai replied:
You're right, idk how I didn't realized that

sipa replied:
You can use runtime asserts in constexpr functions. It just means invocations that would trigger that revert to being runtime evaluated. If the result is assigned to a constexpr variable, it'll error at compile time.

MarcoFalke replied:
Why does c++ not allow constexpr params, so that a static_assert could be used here?

hebasto replied:
:man_shrugging:

sipa replied:
You... could.

```c++
constexpr uint8_t operator ""_u8(unsigned long long byte)
{
    constexpr unsigned long long v = byte;
    static_assert(v < 256);
    return v;
}
```

sipa replied:
C++20 adds (`consteval`)[https://en.cppreference.com/w/cpp/language/consteval] which would be useful here.

MarcoFalke replied:
I'd probably prefer `std::make_array` to be backported. https://en.cppreference.com/w/cpp/experimental/make_array

This would allow to write `MakeArray<uint8_t>(1,2,3)`

hebasto replied:
[`std::to_array`](https://en.cppreference.com/w/cpp/container/array/to_array) from C++20 ?

hebasto commented:
Updated 1f16514614d37edd6e916e0aca731d445101f8f7 -> b6fb4fa999b927dce07bce239532b0f07bd393a2 ([pr20571.01](https://github.com/hebasto/bitcoin/commits/pr20571.01) -> [pr20571.02](https://github.com/hebasto/bitcoin/commits/pr20571.02), [diff](https://github.com/hebasto/bitcoin/compare/pr20571.01..pr20571.02)):

- addressed @elichai's [comment](https://github.com/bitcoin/bitcoin/pull/20571#discussion_r537006012):
> Maybe we should add some overflow assert here? it would be nice if `256_u8` won't compile.

To make compiler proceed checking, all of the `std::array` definitions must be specified with `constexpr`.

hebasto commented:
Closing as this solution seems not future-proof.
